File: mcp-server-golang/tools/converter.go
package tools

import (
	"fmt"
	"mcp-server/types"
)

func GetConverterTools() []types.Tool {
	return []types.Tool{
		{
			Name:        "celsius_to_fahrenheit",
			Description: "Mengkonversi suhu dari Celsius ke Fahrenheit",
			Parameters: map[string]interface{}{
				"type": "object",
				"properties": map[string]interface{}{
					"celsius": map[string]interface{}{
						"type":        "number",
						"description": "Suhu dalam Celsius",
					},
				},
				"required": []string{"celsius"},
			},
			Handler: func(args map[string]interface{}) (interface{}, error) {
				celsius := args["celsius"].(float64)
				fahrenheit := (celsius * 9 / 5) + 32
				kelvin := celsius + 273.15

				fmt.Printf("✅ MCP2 Converter c2f: %.2fC -> %.2fF\n", celsius, fahrenheit)

				return map[string]interface{}{
					"celsius":    celsius,
					"fahrenheit": fahrenheit,
					"kelvin":     kelvin,
				}, nil
			},
		},
		{
			Name:        "fahrenheit_to_celsius",
			Description: "Mengkonversi suhu dari Fahrenheit ke Celsius",
			Parameters: map[string]interface{}{
				"type": "object",
				"properties": map[string]interface{}{
					"fahrenheit": map[string]interface{}{
						"type":        "number",
						"description": "Suhu dalam Fahrenheit",
					},
				},
				"required": []string{"fahrenheit"},
			},
			Handler: func(args map[string]interface{}) (interface{}, error) {
				fahrenheit := args["fahrenheit"].(float64)
				celsius := (fahrenheit - 32) * 5 / 9
				kelvin := celsius + 273.15

				fmt.Printf("✅ MCP2 Converter f2c: %.2fF -> %.2fC\n", fahrenheit, celsius)

				return map[string]interface{}{
					"fahrenheit": fahrenheit,
					"celsius":    celsius,
					"kelvin":     kelvin,
				}, nil
			},
		},
		{
			Name:        "km_to_miles",
			Description: "Mengkonversi jarak dari kilometer ke mil",
			Parameters: map[string]interface{}{
				"type": "object",
				"properties": map[string]interface{}{
					"km": map[string]interface{}{
						"type":        "number",
						"description": "Jarak dalam kilometer",
					},
				},
				"required": []string{"km"},
			},
			Handler: func(args map[string]interface{}) (interface{}, error) {
				km := args["km"].(float64)
				miles := km * 0.621371
				meters := km * 1000
				feet := km * 3280.84

				fmt.Printf("✅ MCP2 Converter km2mil: %.2fkm -> %.2fmil\n", km, miles)

				return map[string]interface{}{
					"km":     km,
					"miles":  miles,
					"meters": meters,
					"feet":   feet,
				}, nil
			},
		},
		{
			Name:        "miles_to_km",
			Description: "Mengkonversi jarak dari mil ke kilometer",
			Parameters: map[string]interface{}{
				"type": "object",
				"properties": map[string]interface{}{
					"miles": map[string]interface{}{
						"type":        "number",
						"description": "Jarak dalam mil",
					},
				},
				"required": []string{"miles"},
			},
			Handler: func(args map[string]interface{}) (interface{}, error) {
				miles := args["miles"].(float64)
				km := miles * 1.60934
				meters := miles * 1609.34
				feet := miles * 5280

				fmt.Printf("✅ MCP2 Converter mil2km: %.2fmil -> %.2fkm\n", miles, km)

				return map[string]interface{}{
					"miles":  miles,
					"km":     km,
					"meters": meters,
					"feet":   feet,
				}, nil
			},
		},
		{
			Name:        "kg_to_pounds",
			Description: "Mengkonversi berat dari kilogram ke pound",
			Parameters: map[string]interface{}{
				"type": "object",
				"properties": map[string]interface{}{
					"kg": map[string]interface{}{
						"type":        "number",
						"description": "Berat dalam kilogram",
					},
				},
				"required": []string{"kg"},
			},
			Handler: func(args map[string]interface{}) (interface{}, error) {
				kg := args["kg"].(float64)
				pounds := kg * 2.20462
				grams := kg * 1000
				ounces := kg * 35.274

				fmt.Printf("✅ MCP2 Converter kg2lb: %.2fkg -> %.2flb\n", kg, pounds)

				return map[string]interface{}{
					"kg":     kg,
					"pounds": pounds,
					"grams":  grams,
					"ounces": ounces,
				}, nil
			},
		},
		{
			Name:        "pounds_to_kg",
			Description: "Mengkonversi berat dari pound ke kilogram",
			Parameters: map[string]interface{}{
				"type": "object",
				"properties": map[string]interface{}{
					"pounds": map[string]interface{}{
						"type":        "number",
						"description": "Berat dalam pound",
					},
				},
				"required": []string{"pounds"},
			},
			Handler: func(args map[string]interface{}) (interface{}, error) {
				pounds := args["pounds"].(float64)
				kg := pounds * 0.453592
				grams := pounds * 453.592
				ounces := pounds * 16

				fmt.Printf("✅ MCP2 Converter lb2kg: %.2flb -> %.2fkg\n", pounds, kg)

				return map[string]interface{}{
					"pounds": pounds,
					"kg":     kg,
					"grams":  grams,
					"ounces": ounces,
				}, nil
			},
		},
	}
}

File: mcp-server-golang/src/tools/converter.ts
import type { Tool } from "../../../types/tool";

export const converterTools: Tool[] = [
  {
    name: "celsius_to_fahrenheit",
    description: "Mengkonversi suhu dari Celsius ke Fahrenheit",
    parameters: {
      type: "object",
      properties: {
        celsius: { type: "number", description: "Suhu dalam Celsius" },
      },
      required: ["celsius"],
    },
    handler: async ({ celsius }: { celsius: number }) => {
      const fahrenheit = (celsius * 9) / 5 + 32;
      console.log(`✅ MCP2 Converter c2f: ${celsius}C -> ${fahrenheit}F`);
      return {
        celsius,
        fahrenheit,
        kelvin: celsius + 273.15,
      };
    },
  },
  {
    name: "fahrenheit_to_celsius",
    description: "Mengkonversi suhu dari Fahrenheit ke Celsius",
    parameters: {
      type: "object",
      properties: {
        fahrenheit: { type: "number", description: "Suhu dalam Fahrenheit" },
      },
      required: ["fahrenheit"],
    },
    handler: async ({ fahrenheit }: { fahrenheit: number }) => {
      const celsius = ((fahrenheit - 32) * 5) / 9;
      console.log(`✅ MCP2 Converter f2c: ${fahrenheit}F -> ${celsius}C`);
      return {
        fahrenheit,
        celsius,
        kelvin: celsius + 273.15,
      };
    },
  },
  {
    name: "km_to_miles",
    description: "Mengkonversi jarak dari kilometer ke mil",
    parameters: {
      type: "object",
      properties: {
        km: { type: "number", description: "Jarak dalam kilometer" },
      },
      required: ["km"],
    },
    handler: async ({ km }: { km: number }) => {
      console.log(`✅ MCP2 Converter km2mil: ${km}km -> ${km * 0.621371}mil`);
      return {
        km,
        miles: km * 0.621371,
        meters: km * 1000,
        feet: km * 3280.84,
      };
    },
  },
  {
    name: "miles_to_km",
    description: "Mengkonversi jarak dari mil ke kilometer",
    parameters: {
      type: "object",
      properties: {
        miles: { type: "number", description: "Jarak dalam mil" },
      },
      required: ["miles"],
    },
    handler: async ({ miles }: { miles: number }) => {
      console.log(
        `✅ MCP2 Converter mil2km: ${miles}mil -> ${miles * 1.60934}km`,
      );
      return {
        miles,
        km: miles * 1.60934,
        meters: miles * 1609.34,
        feet: miles * 5280,
      };
    },
  },
  {
    name: "kg_to_pounds",
    description: "Mengkonversi berat dari kilogram ke pound",
    parameters: {
      type: "object",
      properties: {
        kg: { type: "number", description: "Berat dalam kilogram" },
      },
      required: ["kg"],
    },
    handler: async ({ kg }: { kg: number }) => {
      console.log(`✅ MCP2 Converter kg2lb: ${kg}kg -> ${kg * 2.20462}lb`);
      return {
        kg,
        pounds: kg * 2.20462,
        grams: kg * 1000,
        ounces: kg * 35.274,
      };
    },
  },
  {
    name: "pounds_to_kg",
    description: "Mengkonversi berat dari pound ke kilogram",
    parameters: {
      type: "object",
      properties: {
        pounds: { type: "number", description: "Berat dalam pound" },
      },
      required: ["pounds"],
    },
    handler: async ({ pounds }: { pounds: number }) => {
      console.log(
        `✅ MCP2 Converter lb2kg: ${pounds}lb -> ${pounds * 0.453592}kg`,
      );
      return {
        pounds,
        kg: pounds * 0.453592,
        grams: pounds * 453.592,
        ounces: pounds * 16,
      };
    },
  },
];
